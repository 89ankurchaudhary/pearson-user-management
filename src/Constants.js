export const CUSTOM_STYLES = {
    content : {
      top                   : '50%',
      left                  : '50%',
      right                 : 'auto',
      bottom                : 'auto',
      marginRight           : '-50%',
      transform             : 'translate(-50%, -50%)',
      height                : '100px',
      width                 : '300px'
    }
};

export const INITIAL_USERS = [
    {
      id: 4,
      first_name: "Eve",
      last_name: "Holt",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
    },
    {
      id: 5,
      first_name: "Charles",
      last_name: "Morris",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg"
    },
    {
      id: 6,
      first_name: "Tracey",
      last_name: "Ramos",
      avatar:
        "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg"
    }
  ];

  export const USERS_API = 'https://reqres.in/api/users?page=1&per_page=10';
