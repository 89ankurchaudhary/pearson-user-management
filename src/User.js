import React from "react"

export class User extends React.PureComponent {
    render() {
        const {
            user,
            onOpenDialog,
        } = this.props;

        return (
            <div className="users-thumbnail">
                <img className="users-image" src={user.avatar}></img>
                <div className="users-name">{user.first_name} {user.last_name}</div>
                <div className="users-deletebtn">
                    <a onClick={onOpenDialog}>Delete</a>
                </div>
            </div>
        );
    }
}
