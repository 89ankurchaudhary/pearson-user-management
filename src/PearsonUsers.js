import React, { Component } from "react"
import axios from "axios"
import { Dialog } from './Dialog';
import { User } from './User';
import { INITIAL_USERS, USERS_API } from './Constants';
import { Pagination } from './Pagination';

export class PearsonUsers extends Component {
  constructor(props) {
    super(props);
    this.state = {
      users: INITIAL_USERS,
      modalIsOpen: false,
      selectedUser: null,
      pageNumbers: [],
      firstTimeLoaded: true,
      currentPageNumber: 1
    };
  }

  componentDidMount() {
    this.jsonList(this.state.currentPageNumber);
  }

  removeDuplicateUsers(usersToFilter) {
    const obj = {} , filteredData = [];
    for (let i=0;i < usersToFilter.length; i++)
      obj[usersToFilter[i]['id']] = usersToFilter[i];
    for (var key in obj)
      filteredData.push(obj[key]);
    return filteredData;
  }

  jsonList(pagenumber) {
    axios.get(USERS_API,{
      params: {
        page: pagenumber,
        per_page: 10
      }
    })
    .then((response)=> {
      this.setState((prevState) => {
        let userdata;
        userdata = this.state.firstTimeLoaded ? prevState.users.concat(response.data.data) : response.data.data ;
        const pageNumbers = []; 
        for(let i=1 ; i <= response.data.total_pages ; i++){
          pageNumbers.push(i)
        }
        return {
          users: this.removeDuplicateUsers(userdata),
          pageNumbers: pageNumbers

        };
      });
    })
  }

  openDialog(user) {
    this.setState({
      modalIsOpen: true,
      selectedUser: user,
    });
  }

  closeDialog() {
    this.setState({
      modalIsOpen: false,
    });
  }

  //functionality to delete users 
  deleteUser(){
    const data = this.state.users.filter(i => i.id !== this.state.selectedUser.id)
    this.setState(() => {
      return {users: data};
    })
    this.closeDialog();
  }

  handlePagination(pagenumber) {
    this.setState({
      firstTimeLoaded: false,
    })
    this.jsonList(pagenumber);
  }

  render() {
    return (
      <div className="users">
        <div className="users-header">Pearson User Management</div>
        <div className="users-list">
          { 
            this.state.users.map(user => {
              return (
                <User
                  key={user.id}
                  user={user}
                  onOpenDialog={this.openDialog.bind(this, user)}
                />
              )
            })
          }
        </div>
        <div className="users-pagination">
          {
            this.state.pageNumbers.map((number) => {
              return (
                <Pagination 
                  key={number} 
                  id={number} 
                  handlePagination ={this.handlePagination.bind(this,number)}
                />
              )
            })
          }
        </div>
        <Dialog
          isOpen={this.state.modalIsOpen}
          onDeleteUser={this.deleteUser.bind(this)}
          onCloseDialog={this.closeDialog.bind(this)}
        />
      </div>
    );
  }
}