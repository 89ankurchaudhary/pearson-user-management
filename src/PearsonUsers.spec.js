import React from "react";
import { shallow } from "enzyme";
import { PearsonUsers } from "./PearsonUsers";
import { INITIAL_USERS } from './Constants';
import { User } from './User';

describe("PearsonUsers", () => {
  let component;

  beforeEach(() => {
    PearsonUsers.prototype.jsonList = jest.fn();
    component = shallow(<PearsonUsers />);
    component.setState({
      users: INITIAL_USERS
    });
  });

  it("PersonUsers should render successfully", () => {
    expect(component).toMatchSnapshot();
  });

  it("Header should render properly", () => {
    const userHeader = component.find(".users-header");
    expect(userHeader.text()).toEqual("Pearson User Management");
  });

  it("User should render properly", () => {
    expect(component.find(User).length).toEqual(3);
  });

  it("removeDuplicateUsers should work properly", () => {
    const users = INITIAL_USERS;
    users.concat(INITIAL_USERS[0]);
    expect(component.instance().removeDuplicateUsers(users)).toEqual(INITIAL_USERS);
  });
});
