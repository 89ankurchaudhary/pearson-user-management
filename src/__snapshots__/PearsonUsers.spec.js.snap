// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PearsonUsers PersonUsers should render successfully 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PearsonUsers />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="users-header"
>
          Pearson User Management
</div>,
        <div
          className="users-list"
>
          <User
                    onOpenDialog={[Function]}
                    user={
                              Object {
                                        "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                                        "first_name": "Eve",
                                        "id": 4,
                                        "last_name": "Holt",
                                      }
                    }
          />
          <User
                    onOpenDialog={[Function]}
                    user={
                              Object {
                                        "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                                        "first_name": "Charles",
                                        "id": 5,
                                        "last_name": "Morris",
                                      }
                    }
          />
          <User
                    onOpenDialog={[Function]}
                    user={
                              Object {
                                        "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                                        "first_name": "Tracey",
                                        "id": 6,
                                        "last_name": "Ramos",
                                      }
                    }
          />
</div>,
        <div
          className="users-pagination"
>
          
</div>,
        <Dialog
          isOpen={false}
          onCloseDialog={[Function]}
          onDeleteUser={[Function]}
/>,
      ],
      "className": "users",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Pearson User Management",
          "className": "users-header",
        },
        "ref": null,
        "rendered": "Pearson User Management",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <User
              onOpenDialog={[Function]}
              user={
                            Object {
                                          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                                          "first_name": "Eve",
                                          "id": 4,
                                          "last_name": "Holt",
                                        }
              }
/>,
            <User
              onOpenDialog={[Function]}
              user={
                            Object {
                                          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                                          "first_name": "Charles",
                                          "id": 5,
                                          "last_name": "Morris",
                                        }
              }
/>,
            <User
              onOpenDialog={[Function]}
              user={
                            Object {
                                          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                                          "first_name": "Tracey",
                                          "id": 6,
                                          "last_name": "Ramos",
                                        }
              }
/>,
          ],
          "className": "users-list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "class",
            "props": Object {
              "onOpenDialog": [Function],
              "user": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                "first_name": "Eve",
                "id": 4,
                "last_name": "Holt",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "class",
            "props": Object {
              "onOpenDialog": [Function],
              "user": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                "first_name": "Charles",
                "id": 5,
                "last_name": "Morris",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "6",
            "nodeType": "class",
            "props": Object {
              "onOpenDialog": [Function],
              "user": Object {
                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                "first_name": "Tracey",
                "id": 6,
                "last_name": "Ramos",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "users-pagination",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "isOpen": false,
          "onCloseDialog": [Function],
          "onDeleteUser": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="users-header"
>
            Pearson User Management
</div>,
          <div
            className="users-list"
>
            <User
                        onOpenDialog={[Function]}
                        user={
                                    Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                                                "first_name": "Eve",
                                                "id": 4,
                                                "last_name": "Holt",
                                              }
                        }
            />
            <User
                        onOpenDialog={[Function]}
                        user={
                                    Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                                                "first_name": "Charles",
                                                "id": 5,
                                                "last_name": "Morris",
                                              }
                        }
            />
            <User
                        onOpenDialog={[Function]}
                        user={
                                    Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                                                "first_name": "Tracey",
                                                "id": 6,
                                                "last_name": "Ramos",
                                              }
                        }
            />
</div>,
          <div
            className="users-pagination"
>
            
</div>,
          <Dialog
            isOpen={false}
            onCloseDialog={[Function]}
            onDeleteUser={[Function]}
/>,
        ],
        "className": "users",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Pearson User Management",
            "className": "users-header",
          },
          "ref": null,
          "rendered": "Pearson User Management",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <User
                onOpenDialog={[Function]}
                user={
                                Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                                                "first_name": "Eve",
                                                "id": 4,
                                                "last_name": "Holt",
                                              }
                }
/>,
              <User
                onOpenDialog={[Function]}
                user={
                                Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                                                "first_name": "Charles",
                                                "id": 5,
                                                "last_name": "Morris",
                                              }
                }
/>,
              <User
                onOpenDialog={[Function]}
                user={
                                Object {
                                                "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                                                "first_name": "Tracey",
                                                "id": 6,
                                                "last_name": "Ramos",
                                              }
                }
/>,
            ],
            "className": "users-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "class",
              "props": Object {
                "onOpenDialog": [Function],
                "user": Object {
                  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
                  "first_name": "Eve",
                  "id": 4,
                  "last_name": "Holt",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "class",
              "props": Object {
                "onOpenDialog": [Function],
                "user": Object {
                  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
                  "first_name": "Charles",
                  "id": 5,
                  "last_name": "Morris",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "6",
              "nodeType": "class",
              "props": Object {
                "onOpenDialog": [Function],
                "user": Object {
                  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
                  "first_name": "Tracey",
                  "id": 6,
                  "last_name": "Ramos",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "users-pagination",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "isOpen": false,
            "onCloseDialog": [Function],
            "onDeleteUser": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
